lib/spack/spack/environment/environment.py

```diff
@@ -2500,6 +2461,355 @@ def no_active_environment():
+                item[list_name].remove(user_spec)
+                break
+            except ValueError:
+                pass
+
+        config_dict(self.yaml_content)["definitions"][idx][list_name].remove(user_spec)
+        self.changed = True
+
+    def override_definition(self, user_spec: str, *, override: str, list_name: str) -> None:
+        """Overrides a user spec from an active definition that matches the name passed
+        as argument.
+
+        Args:
+            user_spec: user spec to be overridden
+            override: new spec to be used
+            list_name: name of the definition where to override the spec
+
+        Raises:
+            SpackEnvironmentError: if the user spec cannot be overridden
+        """
+        defs = config_dict(self.pristine_yaml_content).get("definitions", [])
+        msg = f"cannot override {user_spec} with {override} in the '{list_name}' definition"
+
+        for idx, item in self._iterate_on_definitions(defs, list_name=list_name, err_msg=msg):
+            try:
+                sub_index = item[list_name].index(user_spec)
+                item[list_name][sub_index] = override
+                break
+            except ValueError:
+                pass
+
+        config_dict(self.yaml_content)["definitions"][idx][list_name][sub_index] = override
+        self.changed = True
+
+    def _iterate_on_definitions(self, definitions, *, list_name, err_msg):
+        """Iterates on definitions, returning the active ones matching a given name."""
+
+        def extract_name(_item):
+            names = list(x for x in _item if x != "when")
+            assert len(names) == 1, f"more than one name in {_item}"
+            return names[0]
+
+        for idx, item in enumerate(definitions):
+            name = extract_name(item)
+            if name != list_name:
+                continue
+
+            condition_str = item.get("when", "True")
+            if not _eval_conditional(condition_str):
+                continue
+
+            yield idx, item
+        else:
+            raise SpackEnvironmentError(err_msg)
+
+    def set_default_view(self, view: Union[bool, str, pathlib.Path, Dict[str, str]]) -> None:
+        """Sets the default view root in the manifest to the value passed as input.
+
+        Args:
+            view: If the value is a string or a path, it specifies the path to the view. If
+                True the default view is used for the environment, if False there's no view.
+        """
+        if isinstance(view, dict):
+            config_dict(self.pristine_yaml_content)["view"][default_view_name].update(view)
+            config_dict(self.yaml_content)["view"][default_view_name].update(view)
+            self.changed = True
+            return
+
+        if not isinstance(view, bool):
+            view = str(view)
+
+        config_dict(self.pristine_yaml_content)["view"] = view
+        config_dict(self.yaml_content)["view"] = view
+        self.changed = True
```

Comment:
For all these methods that set `self.changed`, shouldn't they confirm that the new data is different than the old before setting it?

Reply:
Not really, for a couple or reasons:
1. Most operations perform an append or a remove, so they don't override existing data
2. (more important) This class performs "lower level" operations for the environment, and it's the environment that checks if the data needs to be updated in the manifest before calling update methods on it. 

For this specific function it would be:

https://github.com/spack/spack/blob/bd86ad12b3353f436224e4c427d0c1017d500405/lib/spack/spack/environment/environment.py#L1563-L1571